### PR TITLE
Bench: system to override opam urls

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Let's test with https://github.com/LPCIC/coq-elpi/pull/672

**Comment by SkySkimmer** [comment]
Doesn't work as opam pin with url replaces the whole opam definition not just the source, so we get the version constraints declared at the url instead of the ones from the package we're trying to modify.

**Comment by SkySkimmer** [comment]
Seems to work

File: dev/bench/bench.sh [hunk 1]
```diff
@@ -368,8 +376,23 @@ create_opam() {
     opam repo add -q --this-switch coq-extra-dev "$OPAM_COQ_DIR/extra-dev"
     opam repo add -q --this-switch coq-released "$OPAM_COQ_DIR/released"
 
-    # Pinning for packages that are not in a repository
-    opam pin add -ynq coq-perennial.dev git+https://github.com/mit-pdos/perennial#coq/tested
```
Comment (silene) [inline dev/bench/bench.sh:372]: Do you intend to move it somewhere else?

Comment (SkySkimmer) [reply dev/bench/bench.sh:372]: We haven't benched perennial for a while so this is just a in-passing cleanup

**Review by silene — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Not sure what's going on, it seems to be picking the wrong version opam file as a base for the pin
ie `opam.OLD/ocaml-OLD/.opam-switch/overlay/coq-elpi/opam` contains `"coq" {>= "8.19" & < "8.21"}` unlike the .dev package which has = dev.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by silene** [comment]
Perhaps Opam is choosing the file from the latest remote? (I see that the script first adds `extra-dev` and then `released`, which is the opposite of my own setup.)

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by silene** [comment]
I have tested locally; switching the order does not make a difference, unfortunately. However, `opam pin add -e` does respect the version specification, that is, `opam pin add -e coq-foo.version` forces the use of the given version as the basis for the pin.

**Comment by SkySkimmer** [comment]
The opam show prints what we want, but then it fails

**Comment by silene** [comment]
I am confused. Why is there already an error message before the line `sorted_coq_opam_packages = coq-elpi coq-bignums coq-mathcomp-ssreflect` has even be printed?

**Comment by SkySkimmer** [comment]
probably the opam install --show-actions from https://github.com/coq/coq/blob/master/dev/bench/sort-by-deps.sh

**Comment by SkySkimmer** [comment]
I can reproduce locally
~~~bash
# pin coq to .dev, then
env OPAMEDITOR="echo 'url { src: \"git+https://github.com/skyskimmer/coq-elpi#old-elpi\" }' >> " opam pin add -nye coq-elpi.dev - </dev/null
opam install --show-actions coq-elpi
# opam not happy
~~~
It looks like we need --skip-updates.

Bench machines are busy right now so we will see tomorrow.

**Comment by SkySkimmer** [comment]
@coqbot bench

File: dev/bench/bench.sh [hunk 2]
```diff
@@ -78,9 +80,14 @@ check_variable () {
 : "${old_coq_version:=dev}"
 : "${num_of_iterations:=1}"
 : "${timeout:=3h}"
-: "${coq_opam_packages:=coq-test-suite coq-bignums coq-hott coq-performance-tests-lite coq-engine-bench-lite coq-mathcomp-ssreflect coq-mathcomp-fingroup coq-mathcomp-algebra coq-mathcomp-solvable coq-mathcomp-field coq-mathcomp-character coq-mathcomp-odd-order coq-math-classes coq-corn coq-compcert coq-equations coq-metacoq-utils coq-metacoq-common coq-metacoq-template coq-metacoq-pcuic coq-metacoq-safechecker coq-metacoq-erasure coq-metacoq-translations coq-color coq-coqprime coq-coqutil coq-bedrock2 coq-rewriter coq-fiat-core coq-fiat-parsers coq-fiat-crypto-with-bedrock coq-unimath coq-coquelicot coq-iris-examples coq-verdi coq-verdi-raft coq-fourcolor coq-rewriter-perf-SuperFast coq-vst coq-category-theory coq-neural-net-interp-computed-lite}"
+: "${coq_opam_packages:=coq-elpi coq-bignums coq-mathcomp-ssreflect}" # coq-mathcomp-fingroup coq-mathcomp-algebra coq-mathcomp-solvable coq-mathcomp-field coq-mathcomp-character coq-mathcomp-odd-order coq-mathcomp-finmap coq-mathcomp-classical coq-mathcomp-bigenough coq-mathcomp-analysis
```
Comment (SkySkimmer) [inline dev/bench/bench.sh:83]:
```suggestion
: "${coq_opam_packages:=coq-elpi coq-bignums coq-mathcomp-ssreflect coq-mathcomp-fingroup coq-mathcomp-algebra coq-mathcomp-solvable coq-mathcomp-field coq-mathcomp-character coq-mathcomp-odd-order coq-mathcomp-finmap coq-mathcomp-classical coq-mathcomp-bigenough coq-mathcomp-analysis}"
```

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Should be ready

**Comment by SkySkimmer** [comment]
The versioned package name and the url you want it to use

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Comment by SkySkimmer** [comment]
Seems like the bot forgot to merge? @coq/coqbot-maintainers

**Comment by ppedrot** [comment]
@coqbot merge now or else!

**Pull request merged**